From bd3c1dda5f7cc9f62dbcdd9fc8fa037f3fc7d2e1 Mon Sep 17 00:00:00 2001
From: Karthikeyan Kathirvel <kathirve@codeaurora.org>
Date: Tue, 24 Aug 2021 17:02:36 +0530
Subject: [PATCH] ath11k: mon ring interrupt changes

In ahb ppdu-end-interrupts are used as an interrupt line for monitor
status ring, ppdu end interrupts are not available in PCI chipset and
require srng msi interrupts to reap monitor status ring.

Interrupt threshold setting for monitor status interrupt was timingly
interrupted, so the dependent ppdu stats info from monitor status ring
is quite old. Set interrupt batch counter threshold as 4 for monitor
status ring to get the frequent interrupts rather than depending on
timer interrupt alone.

Signed-off-by: Karthikeyan Kathirvel <kathirve@codeaurora.org>
---
 drivers/net/wireless/ath/ath11k/dp.c | 27 ++++++++++++++++++++++++++-
 1 file changed, 26 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath11k/dp.c
+++ b/drivers/net/wireless/ath/ath11k/dp.c
@@ -313,8 +313,18 @@ int ath11k_dp_srng_setup(struct ath11k_b
 	case HAL_RXDMA_MONITOR_STATUS:
 		params.low_threshold = num_entries >> 3;
 		params.flags |= HAL_SRNG_FLAGS_LOW_THRESH_INTR_EN;
-		params.intr_batch_cntr_thres_entries = 0;
 		params.intr_timer_thres_us = HAL_SRNG_INT_TIMER_THRESHOLD_RX;
+		/* In case of PCI chipsets, we dont have PPDU end interrupts,
+		 * so MONITOR STATUS ring is reaped by receiving MSI from srng.
+		 * Keep batch threshold as 8 so that an interrupt is received for
+		 * every 4 frames in MONITOR_STATUS ring.
+		 */
+		if ((type == HAL_RXDMA_MONITOR_STATUS) &&
+				(params.flags & HAL_SRNG_FLAGS_MSI_INTR))
+			params.intr_batch_cntr_thres_entries = 4;
+		else
+			params.intr_batch_cntr_thres_entries = 0;
+
 		break;
 	case HAL_WBM2SW_RELEASE:
 		if (ring_num < 3) {
